### PR TITLE
Particle flow identification of electrons in extended eta

### PR DESCRIPTION
**Pull request by a-kapoor**
#### PR description:
The reconstruction efficiency of electrons in |&eta;|>2.5 was significantly improved via [#35309](https://github.com/cms-sw/cmssw/pull/35309). These electrons are now available as gsf electrons but were not allowed to enter Particle Flow (PF) when PF ID for electrons in |&eta;|<=2.5 was added via [#35403](https://github.com/cms-sw/cmssw/pull/35403), [#36243](https://github.com/cms-sw/cmssw/pull/36243/), [#36292](https://github.com/cms-sw/cmssw/pull/36292) 

This PR adds the code needed for the PF ID of electrons in |&eta;|>2.5. Dedicated PF IDs have been developed, for electrons in this region. There are two dedicated trainings, one for  **Region 1:** 2.5<|&eta;|<=2.65 and one for **Region 2:** |&eta;|>2.65, given that electrons in these regions look very different. More details can be found in this [presentation](https://indico.cern.ch/event/1146227/#10-pf-egm-id-of-eta-extended-e).

```mermaid
graph TD
A(GSF Electrons) -->|PF-Electron-ID in PFEGammaFilters| B("PF Electrons")
PF-Electron-ID -->|"|&eta;|<=2.5 "| C[ID already in CMSSW]
PF-Electron-ID -->|" 2.5<|&eta;|<=2.65"| D["extended eta region 1 ID (**This PR**)"]
PF-Electron-ID -->|"|&eta;|>2.65"| E["extended eta region 2 ID (**This PR**)"]
```

This PR depends on [#25](https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/25) to cms-data that has been made to merge the model files.

Since before this PR, electrons in |&eta;|>2.5 must have been identified by Particle Flow as other objects like photons and jets, this PR might lead to some efficiency changes for primarily PF Photons and to some extent Taus and Jets in |&eta;|>2.5.

Some points:
It is understood that this ID will not give the most optimal performance that can be achieved in this region, since some kinds of electrons needed for the training were not fully produced due to hard cuts on | &eta; | in the Tau Gun and QCD_bcToE samples. The hard cuts for the TauGun sample were in the config so this can be easily fixed in the next iteration of samples. However, for the QCD_bcToE sample the cut was hard-coded in [BCToEFilterAlgo.cc](https://github.com/cms-sw/cmssw/blob/d8184188a4f5843b1b4a84e5d800ad178ae8bacf/GeneratorInterface/GenFilters/plugins/BCToEFilterAlgo.cc#L13). Given these constraints, the model in |&eta;|>2.65 has only 3 output nodes instead of the 5 nodes that all other DNN models have. In [#37869](https://github.com/cms-sw/cmssw/pull/37869) we have fixed this. Given the timeline for 12_4_0, update of samples is not possible in this iteration. The PF ID of electrons in |&eta;|>2.5 will thus be updated in a later release, possibly at the end of this year. 

**Main code changes:**
- Output dimension of the DNN for each region is now configurable independently by changing it from a "single int" to a "vector of ints"
- PFEGammaFilters.cc has been modified to allow the new electrons to enter PF and also select the right ID thresholds for these electrons.
- ElectronDNNEstimator has been modified to select the right model depending on the pt and &eta; of the electron.

#### PR validation:

A general validation to check the effect on objects downstream in PF is currently ongoing. A dedicated validation to check the effect on PF Taus is also being carried out. This section will be updated as the results of the validation become available.

### CODE REVIEW

**Comment by jpata** [comment]
test parameters:
 - pull_requests: https://github.com/https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/25

**Comment by jpata** [comment]
type egamma

**Comment by jpata** [comment]
test parameters:
 - pull_requests = https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/25

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
Hi @a-kapoor , during your talk at the [Joint PF and EGM meeting](https://indico.cern.ch/event/1146227/#10-pf-egm-id-of-eta-extended-e), there has been some discussion about performing a validation of the extended PFID. Are these results available?

**Comment by a-kapoor** [comment]
> Hi @a-kapoor , during your talk at the [Joint PF and EGM meeting](https://indico.cern.ch/event/1146227/#10-pf-egm-id-of-eta-extended-e), there has been some discussion about performing a validation of the extended PFID. Are these results available?

Hi @clacaputo, a general validation and a dedicated PF Tau validation are ongoing. We do not have the results yet, but we will update the thread once we have something.

**Comment by swagata87** [comment]
@cmsbuild please test

**Review by clacaputo — COMMENTED** [review]
Hi @a-kapoor , I've left a few comments on the code.

About the comparison, there are lots of minor differences. For examples

**all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFCandidates_particleFlow__reRECO_obj_eta** where there is a decrease of the number of entries
all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFCandidates_particleFlow__reRECO_obj_eta

or **all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFJets_ak4PFJets__reRECO_obj_eta** where the eta of a jet in the range `[-1.5,-1]` shift a bit
all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFJets_ak4PFJets__reRECO_obj_eta

Do you know if these behaviour are expected? I can't understand the decrease in the entries for `recoPFCandidates`

File: RecoEgamma/ElectronIdentification/src/ElectronDNNEstimator.cc [hunk 1]
```diff
@@ -69,6 +80,8 @@ const std::vector<std::string> ElectronDNNEstimator::dnnAvaibleInputs = {
      "full5x5_e1x5_ratio_full5x5_e5x5",
      "full5x5_r9",
      "gsfTrack.trackerLayersWithMeasurement",
+     "gsfTrack.numberOfValidPixelBarrelHits",
```
[the remark this answers is not in the document]
Comment (a-kapoor) [reply RecoEgamma/ElectronIdentification/src/ElectronDNNEstimator.cc:83]: I am not sure if I fully understand what you meant by old models here as the only models introduced by this PR are in regions where there were no old models **(|eta|>2.5)**. Any model (old or new) can use either this complete list or a subset of this list when deploying using [ElectronDNNEstimator](https://github.com/cms-sw/cmssw/pull/37880/files/3e477834e2e2c7812bd219f044f1017845e70a84#diff-c61484fdc9801bba3b67585f6d573de4072def2b0dfb3681fff864c68b55f24c) without changing anything in the code. The exact list of variables that any model will be using is defined in a separate text file.   

So, no. It should not create any problems.  As an example, the currently merged models in regions with **|eta|<2.5** do not even use these two new variables. Still, they are deployed using the same [ElectronDNNEstimator](https://github.com/cms-sw/cmssw/pull/37880/files/3e477834e2e2c7812bd219f044f1017845e70a84#diff-c61484fdc9801bba3b67585f6d573de4072def2b0dfb3681fff864c68b55f24c).

File: RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc [hunk 2]
```diff
@@ -498,18 +492,24 @@ void PFEGammaFilters::fillPSetDescription(edm::ParameterSetDescription& iDesc) {
   iDesc.add<double>("electron_maxElePtForOnlyMVAPresel", 50.0);
   iDesc.add<bool>("allowEEEinPF", false);
```
[the remark this answers is not in the document]
Comment (a-kapoor) [reply RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:493]: You are right. This is not needed. I just removed it.

**Comment by a-kapoor** [comment]
> Hi @a-kapoor , I've left a few comments on the code.
> 
> About the comparison, there are lots of minor differences. For examples
> 
> **all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFCandidates_particleFlow__reRECO_obj_eta** where there is a decrease of the number of entries all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFCandidates_particleFlow__reRECO_obj_eta
> 
> or **all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFJets_ak4PFJets__reRECO_obj_eta** where the eta of a jet in the range `[-1.5,-1]` shift a bit all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFJets_ak4PFJets__reRECO_obj_eta
> 
> Do you know if these behaviour are expected? I can't understand the decrease in the entries for `recoPFCandidates`

**Comment by swagata87** [comment]
@cmsbuild please test

**Comment by swagata87** [comment]
urgent

Aiming for 12_4_0_pre4. It would be great if it can go in and new MC samples of 12_4_X can be made with this development included. This will allow JME to check its interplay with PUPPI etc. 
There are 2 talks planned on this PR, if people want to further discuss this. 
 - one is today in PPD https://indico.cern.ch/event/1158486/#30-retraining-of-pf-electron-i
 - another is tomorrow in RECO https://indico.cern.ch/event/1158558/#7-egamma-pfid-for-eta-extended

**Comment by lfinco** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
About https://github.com/cms-sw/cmssw/pull/37880/commits/48ce095040a34be6d505c9cdd5d09f629102499c , wouldn't be better to do something like this: 

```python
def customiseFor37756(process):
    """https://github.com/cms-sw/cmssw/pull/37756
    Removal of use_preshower parameter from PFECALSuperClusterProducer
    """
    for prod in producers_by_type(process, 'PFECALSuperClusterProducer'):
        if hasattr(prod, 'use_preshower'):
            delattr(prod, 'use_preshower')
```

https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/customizeHLTforCMSSW.py#L172-L178

As was done in #37756 ? But maybe @cms-sw/hlt-l2 can confirm

**Comment by swagata87** [comment]
> As was done in #37756 ? But maybe @cms-sw/hlt-l2 can confirm

I think that we use `customizeHLTforCMSSW.py` to make change in current HLT menu, while Anshul changed in Phase2 HLT menu, which need to be changed directly as he did. But yes, Marino and Martin can confirm.

**Comment by Martin-Grunewald** [comment]
Yes, the Phase-II menu (only) needs to be modified directly as it is not a dump from a source maintained in ConfDb.

**Comment by clacaputo** [comment]
Hi @a-kapoor , @cms-sw/egamma-pog-l2 , while waiting for the full validation, it could be good to run wf `11834` with more stat (~5k-10k events) locally and run a DQM comparison [1]. This should give and hint on the possible difference that this PR can introduce in downstream quantities and should be feasible in a reasonable amount of time. 
What do you think? It would be great to have this by tomorrow's reco meeting.

[1] https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_outputs_DQM_style

**Comment by swagata87** [comment]
> Hi @a-kapoor , @cms-sw/egamma-pog-l2 , while waiting for the full validation, it could be good to run wf `11834` with more stat (~5k-10k events) locally and run a DQM comparison [1]. This should give and hint on the possible difference that this PR can introduce in downstream quantities and should be feasible in a reasonable amount of time. What do you think? It would be great to have this by tomorrow's reco meeting.
> 
> [1] https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_outputs_DQM_style

ok we can try. 
Just for my understanding, this can't be done by the `bot` right? 
the `high_stats` option won't work here I guess.. for several thousand events?

**Comment by missirol** [comment]
+hlt

(the Phase-2 HLT menu now works, and the change in it is, iiuc, unconsequential)

**Comment by clacaputo** [comment]
> ok we can try. Just for my understanding, this can't be done by the `bot` right? the `high_stats` option won't work here I guess.. for several thousand events?

Unfortunately, it is something you have to do manually setting up a base_release and base+PR, running for both of them the wf, and then running the validateDQM script

**Comment by a-kapoor** [comment]
@jpata @clacaputo
Given the time constraints, we are already sharing plots that have just arrived. @akanugan has produced a **Reference** vs **Reference+PR** comparison in QCD (no pileup) events. We do not see any significant changes in objects downstream. Below are some plots (all PF objects, and PF electrons). 
image

As expected some fake electrons have appeared in the region |η|>2.5. This is expected because, earlier, this region was not open to PF electrons at all. We will also see signal electrons appear in this region when we see Z→ee validation plots, for example.

You can find the full list of plots [here](https://akanugan.web.cern.ch/akanugan/tmp_EGMID/)!
We will keep you updated as more results are available.

**Comment by a-kapoor** [comment]
Dear @jpata, @clacaputo 

We now have validation plots for Z→ee as well. Thanks to @akanugan 
image
image



Changes are as expected and in the right direction:
- More signal electrons at |η|>2.5
- new signal electrons at all φ
- Slightly fewer photons in |η|>2.5 
- Expected changes in jets (2.5<|η|<3.0) 
- Slightly better jet pT resolution at high eta
- No change in PF objects overall

Please let us know if it would be possible to now sign off the PR for 12_4_0.

[Link to all the plots](https://akanugan.web.cern.ch/akanugan/PF_ExtEleID_new_afterBugFix/)!

**Comment by clacaputo** [comment]
+reconstruction

- PF ID for electron extended above |eta| > 2.5
- New models available in https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/25
- Reco changes are in line with the PR
  - validation plots QCD noPU: https://github.com/cms-sw/cmssw/pull/37880#issuecomment-1126260849
  - validation plots Zee: https://github.com/cms-sw/cmssw/pull/37880#issuecomment-1126668267

**Comment by perrotta** [comment]
+1

**Pull request merged**